File: Binary Search/C++/GFG - BINARY SERACH.cpp
// GFG - Binary Search
// QUESTION LINK - https://www.geeksforgeeks.org/problems/binary-search-1587115620/1

class Solution {
  public:
    int binarysearch(vector<int> &arr, int k) {
        int low = 0;
        int high = arr.size() - 1;
        int ans = -1;

        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (arr[mid] == k) {
                ans = mid;
                high = mid - 1;
            } else if (arr[mid] > k) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }

        return ans;
    }
};

File: Binary Search/Python/GFG - BINARY SERACH.py
# GFG - Binary Search
# QUESTION LINK - https://www.geeksforgeeks.org/problems/binary-search-1587115620/1

class Solution:
    def binarysearch(self, arr, k):
        low = 0
        high = len(arr) - 1
        ans = -1

        while low <= high:
            mid = low + (high - low) // 2
            if arr[mid] == k:
                ans = mid
                high = mid - 1  # For first occurrence
            elif arr[mid] > k:
                high = mid - 1
            else:
                low = mid + 1

        return ans
